### PR TITLE
[FLINK-34634]Fix that Restarting the job will not read the changelog anymore if it stops before the synchronization of meta information is complete and some table is removed

### PR DESCRIPTION
**Pull request by loserwang1024**
### What's the problem
Once, I removed a table from the option and then restarted the job from the savepoint, but the job couldn't read the binlog anymore. When I checked the logs, I found an Error level log stating: ' The enumerator received invalid request meta group id 6, the valid meta group id range is [0, 4].'

It appears that the Reader is requesting more splits than the Enumerator is aware of.
However, the code should indeed remove redundant split information from the Reader as seen in https://github.com/ververica/flink-cdc-connectors/pull/2292. So why does this issue occur?

### why occurs
image
Upon examining the code, I discovered the cause. If the job stops before completing all the split meta information and then restarts, this issue occurs. Suppose that the totalFinishedSplitSize of binlogSplit in the Reader is 6, and no meta information has been synchronized, leaving the finishedSnapshotSplitInfos of binlogSplit in the Reader empty. After restarting, the totalFinishedSplitSize of binlogSplit in the Reader equals (6 - (0 - 0)) which is still 6, but in the Enumerator, it is only 4(the removed table have two split). This could lead to an out-of-range request.
image
### How to reproduce

-  Add Thread.sleep(1000L) in com.ververica.cdc.connectors.mysql.source.reader.MySqlSourceReader#handleSourceEvents to postpone split meta infos synchronization.

```java
public void handleSourceEvents(SourceEvent sourceEvent) {
else if (sourceEvent instanceof BinlogSplitMetaEvent) {
    LOG.debug(
            "Source reader {} receives binlog meta with group id {}.",
            subtaskId,
            ((BinlogSplitMetaEvent) sourceEvent).getMetaGroupId());
    try {
        Thread.sleep(1000L);
    } catch (InterruptedException e) {
        throw new RuntimeException(e);
    }
    fillMetadataForBinlogSplit((BinlogSplitMetaEvent) sourceEvent);
} 
```

- Add Thread.sleep(500L) in com.ververica.cdc.connectors.mysql.source.NewlyAddedTableITCase#testRemoveTablesOneByOne to trigger savepoint before meta infos synchronization finishes.
```java
// step 2: execute insert and trigger savepoint with all tables added
{
    // ..ingore 

    waitForSinkSize("sink", fetchedDataList.size());
    Thread.sleep(500L);
    assertEqualsInAnyOrder(fetchedDataList, TestValuesTableFactory.getRawResults("sink"));
    finishedSavePointPath = triggerSavepointWithRetry(jobClient, savepointDirectory);
    jobClient.cancel().get();
}

// test removing table one by one, note that there should be at least one table remaining
for (int round = 0; round < captureAddressTables.length - 1; round++) {
...
}
```

- Add chunk-meta.group.size  =2 in com.ververica.cdc.connectors.mysql.source.NewlyAddedTableITCase#getCreateTableStatement

Finally, run test(com.ververica.cdc.connectors.mysql.source.NewlyAddedTableITCase#testJobManagerFailoverForRemoveTable), the error log will occur.

### CODE REVIEW

**Comment by loserwang1024** [comment]
@ruanhang1993 , @leonardBang , @PatrickRen , CC

**Comment by Jiabao-Sun** [comment]
Thanks @loserwang1024 for investigating this issue. 

IMHO, I think this fix is a bit complicated. 
Can we simply consider that in the previous PR, the split that has not yet completed the meta info synchronization should not be deleted? 

@ruanhang1993  what do you think ?

**Comment by loserwang1024** [comment]
> the split that has not yet completed the meta info synchronization should not be deleted?

What if the snapshot splits are not completed, and then removed from tableList? The meta info won't be removed later , and  also won't be finished , thus the job won't come into stream phase .

**Comment by Jiabao-Sun** [comment]
@loserwang1024 Please correct me if I'm wrong. 
In PR-2292, only unnecessary finished snapshot splits were removed from `BinlogSplit`, which means the checkpoint is already in the stream phase. As for the checkpoint in the snapshot phase, it should not be affected by PR-2292.

**Review by Jiabao-Sun — APPROVED** [review]
Thanks for clarifying, it makes sense to me.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java [hunk 1]
```diff
@@ -298,22 +298,35 @@ private void sendBinlogMeta(int subTask, BinlogSplitMetaRequestEvent requestEven
                             finishedSnapshotSplitInfos, sourceConfig.getSplitMetaGroupSize());
         }
         final int requestMetaGroupId = requestEvent.getRequestMetaGroupId();
-
-        if (binlogSplitMeta.size() > requestMetaGroupId) {
+        final int totalFinishedSplitSizeOfReader = requestEvent.getTotalFinishedSplitSize();
+        final int totalFinishedSplitSizeOfEnumerator = splitAssigner.getFinishedSplitInfos().size();
+        if (totalFinishedSplitSizeOfReader > totalFinishedSplitSizeOfEnumerator) {
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java:303]: Add some logs here.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/reader/MySqlSourceReader.java [hunk 2]
```diff
@@ -395,11 +396,17 @@ private void fillMetadataForBinlogSplit(BinlogSplitMetaEvent metadataEvent) {
         MySqlBinlogSplit binlogSplit = uncompletedBinlogSplits.get(metadataEvent.getSplitId());
         if (binlogSplit != null) {
             final int receivedMetaGroupId = metadataEvent.getMetaGroupId();
+            final int receivedTotalFinishedSplitSize = metadataEvent.getTotalFinishedSplitSize();
             final int expectedMetaGroupId =
                     ChunkUtils.getNextMetaGroupId(
                             binlogSplit.getFinishedSnapshotSplitInfos().size(),
                             sourceConfig.getSplitMetaGroupSize());
-            if (receivedMetaGroupId == expectedMetaGroupId) {
+            if (receivedTotalFinishedSplitSize < binlogSplit.getTotalFinishedSplitSize()) {
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/reader/MySqlSourceReader.java:404]: Add some logs here.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/source/enumerator/IncrementalSourceEnumerator.java [hunk 3]
```diff
@@ -307,8 +307,17 @@ private void sendStreamMetaRequestEvent(int subTask, StreamSplitMetaRequestEvent
                             finishedSnapshotSplitInfos, sourceConfig.getSplitMetaGroupSize());
         }
         final int requestMetaGroupId = requestEvent.getRequestMetaGroupId();
-
-        if (finishedSnapshotSplitMeta.size() > requestMetaGroupId) {
+        final int totalFinishedSplitSizeOfReader = requestEvent.getTotalFinishedSplitSize();
+        final int totalFinishedSplitSizeOfEnumerator = splitAssigner.getFinishedSplitInfos().size();
+        if (totalFinishedSplitSizeOfReader > totalFinishedSplitSizeOfEnumerator) {
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/source/enumerator/IncrementalSourceEnumerator.java:312]: Add logs.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/source/meta/events/StreamSplitMetaEvent.java [hunk 4]
```diff
@@ -43,10 +43,14 @@ public class StreamSplitMetaEvent implements SourceEvent {
      */
     private final List<byte[]> metaGroup;
 
-    public StreamSplitMetaEvent(String splitId, int metaGroupId, List<byte[]> metaGroup) {
+    private final int totalFinishedSplitSize;
+
+    public StreamSplitMetaEvent(
+            String splitId, int metaGroupId, List<byte[]> metaGroup, int totalFinishedSplitSize) {
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/source/meta/events/StreamSplitMetaEvent.java:49]: Add `@Nullable`

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/events/BinlogSplitMetaEvent.java [hunk 5]
```diff
@@ -43,10 +43,14 @@ public class BinlogSplitMetaEvent implements SourceEvent {
      */
     private final List<byte[]> metaGroup;
 
-    public BinlogSplitMetaEvent(String splitId, int metaGroupId, List<byte[]> metaGroup) {
+    private final int totalFinishedSplitSize;
+
+    public BinlogSplitMetaEvent(
+            String splitId, int metaGroupId, List<byte[]> metaGroup, int totalFinishedSplitSize) {
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/events/BinlogSplitMetaEvent.java:49]: Add `@Nullable`

Comment (loserwang1024) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/events/BinlogSplitMetaEvent.java]: Done it.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/source/meta/events/StreamSplitMetaEvent.java (hunk 4 above)
Comment (loserwang1024) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/source/meta/events/StreamSplitMetaEvent.java]: Done it

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java (hunk 1 above)
Comment (loserwang1024) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/source/enumerator/MySqlSourceEnumerator.java:303]: done it

**Pull request merged**